sympy/combinatorics/perm_groups.py

```diff
@@ -3219,17 +3219,15 @@ def is_cyclic(self):
                 self._is_abelian = True
                 return True
 
-        for p in factors:
-            pgens = []
-            for g in self.generators:
-                pgens.append(g**p)
-            if self.index(self.subgroup(pgens)) != p:
-                self._is_cyclic = False
-                return False
+        if not self.is_abelian:
+            self._is_cyclic = False
+            return False
 
-        self._is_cyclic = True
-        self._is_abelian = True
-        return True
+        self._is_cyclic = all(
+            any(g**(order//p) != self.identity for g in self.generators)
+            for p in factors
```

Comment:
This can be made more efficient by adding `if factors[p] > 1`. If `factors[p] == 1`, then the p-component is always cyclic.